File: fixed_code.js
const { Product, Inventory, sequelize } = require('../models');

// Safely creates a new product and its initial inventory.
const createProduct = async (req, res) => {
  try {
    const { name, sku, price, warehouse_id, initial_quantity } = req.body;
    
    // 1. Check for required fields.
    if (!name || !sku || !price || !warehouse_id) {
      return res.status(400).json({
        success: false,
        message: 'Missing required fields: name, sku, price, and warehouse_id are all required.'
      });
    }

    // 2. Check for a duplicate SKU.
    const existingProduct = await Product.findOne({ where: { sku } });
    if (existingProduct) {
      return res.status(409).json({
        success: false,
        message: `A product with SKU '${sku}' already exists.`
      });
    }

    // 3. Create both records in a transaction (all or nothing).
    const result = await sequelize.transaction(async (t) => {
      const product = await Product.create({
        name, sku, price: parseFloat(price)
      }, { transaction: t });

      const inventory = await Inventory.create({
        product_id: product.id,
        warehouse_id: parseInt(warehouse_id),
        quantity: parseInt(initial_quantity) || 0
      }, { transaction: t });

      return { product, inventory };
    });

    // 4. Success! Send a 201 Created response.
    res.status(201).json({
      success: true,
      message: 'Product created successfully',
      product_id: result.product.id
    });

  } catch (error) {
    // If anything fails, log the error and send a generic 500 response.
    console.error('Error creating product:', error);
    res.status(500).json({
      success: false,
      message: 'An internal server error occurred.'
    });
  }
};

module.exports = { createProduct };

File: fixed_code.py
from flask import request, jsonify
from app import app, db
from models import Product, Inventory
@app.route('/api/products', methods=['POST'])
def create_product():
    data = request.get_json()

    # 1. Validate required fields
    if not data:
        return jsonify({"error": "No data provided"}), 400

    required_fields = ['name', 'sku', 'price', 'warehouse_id', 'initial_quantity']
    for field in required_fields:
        if field not in data: 
            return jsonify({"error": f"Missing required field: {field}"}), 400

    # 2. Check if SKU is unique
    if Product.query.filter_by(sku=data['sku']).first():
        return jsonify({"error": "A product with this SKU already exists."}), 409

    # 3. Use transaction for consistency
    try:
        product = Product(
            name=data['name'],
            sku=data['sku'],
            price=float(data['price'])
        )
        db.session.add(product)
        db.session.flush()

        inventory = Inventory(
            product_id=product.id,
            warehouse_id=data['warehouse_id'],
            quantity=int(data['initial_quantity'])
        )
        db.session.add(inventory)
        db.session.commit()

        return jsonify({
            "message": "Product created successfully",
            "product_id": product.id
        }), 201

    except Exception as e:
        db.session.rollback()
        app.logger.error(f"Error creating product: {e}")
        return jsonify({"error": "Internal server error"}), 500